api/src/main/java/jakarta/enterprise/invoke/Invokable.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright (c) 2022 Red Hat and others
+ *
+ * This program and the accompanying materials are made available under the
+ * Apache Software License 2.0 which is available at:
+ * https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package jakarta.enterprise.invoke;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Marks a method as <em>invokable</em>. For each invokable method belonging to
+ * a managed bean class, the CDI container allows obtaining an {@link Invoker}
+ * through which the method can be invoked.
+ * <p>
+ * When this annotation is present on a class or interface, it marks all non-private
+ * methods (except constructors) declared in this class as invokable. Note that this
+ * does not apply to methods declared in superclasses or superinterfaces.
+ * <p>
+ * Attempting to mark a private method or a constructor of a managed bean class
+ * as invokable is a definition error.
+ * <p>
+ * Methods that are marked as invokable but do not belong to managed beans are ignored.
+ * <p>
+ * Note that multiple managed beans may inherit an invokable method from a common
+ * supertype. In that case, each bean conceptually has its own invokable method
+ * and for example an invoker obtained for one bean cannot be used to invoke
+ * the method on the other bean.
```

Comment:
I don't think this isolation makes much sense.
You can have something like class `Foo` with method `ping()` and then you can have `n` implementations (`FooImpl1`, `FooImpl2`, ...) of this class that are beans.
Some of the impls you have invokers for might be disabled, specialized, vetoed, ....
You'd need to create and keep reference to a separate invoker for each variant of the bean - despite the fact that you don't care about what else impl variants bring to the table, in fact you might not even care which impl it is, you just want to get the default one that comes via `@Inject Foo` and execute the method on that one.
Determining which invoker you can use at runtime could become a hassle.

I'd rather have an invoker bound to a method and so long as you pass in a "compatible" bean which has such method, you invoke it there.

Reply:
Good question, conceptually. I probably had more reasons for it, which I don't remember exactly right now, but one is fairly obvious: if I want an invoker to be able to automatically lookup the target instance (`InvokerBuilder.setInstanceLookup()`), then the invoker must be bound not only to the method, but also to the _bean_.

Reply:
Fair enough but you get the invoker from a `ProcessManagedBean` even where you can see bean types (and qualifiers). You can use those for lookup regardless of what the resulting bean is - actually that's exactly what will happen anyway if you have an alternative in place; you'll end up invoking it on different impl instance.

On the topic of instance lookup, assuming the rules weren't this strict, I was thinking if the method could be something like `InvokerBuilder.setInstanceLookup(TypeLiteral<T> t, Annotation... a)` allowing you to specify how to perform the lookup while still leaving dependent instance handling to CDI. Obviously this brings the issue of assignability but that's the same as with many other parameters user can wrongly pass in.

Reply:
Good point about alternatives -- I don't think the existing spec text (in the javadoc) mentions it, but I think we should only discover invokable methods on enabled beans, or something like that.

Reply:
Well, `ProcessBean` (or managed bean in this case) is fired for enabled beans only, so that's automatically true, at least via portable extensions. However IIRC, that doesn't solve the alternative issue fully as you can have multiple enabled alternatives (meaning alternatives selected for given archive) for `Foo`, each with a different priority.